File: event-sdk-generator/packages/runtime-typescript/src/middleware/logging.ts
import type { TransportAdapter, EventEnvelope, PublishResult, Logger } from "../types.js";

const consoleLogger: Logger = {
  info: (message, context) => console.log(JSON.stringify({ level: "info", message, ...context })),
  error: (message, context) => console.error(JSON.stringify({ level: "error", message, ...context })),
};

export function withLogging(adapter: TransportAdapter, logger: Logger = consoleLogger): TransportAdapter {
  return {
    async publish(envelope: EventEnvelope): Promise<PublishResult> {
      const start = Date.now();
      logger.info("publishing event", { eventId: envelope.eventId, version: envelope.version, traceId: envelope.metadata.traceId });
      try {
        const result = await adapter.publish(envelope);
        logger.info("event published", { eventId: envelope.eventId, messageId: result.messageId, durationMs: Date.now() - start });
        return result;
      } catch (err) {
        logger.error("event publish failed", { eventId: envelope.eventId, error: String(err), durationMs: Date.now() - start });
        throw err;
      }
    },
  };
}

File: packages/runtime-go/middleware/logging.go
package middleware

import (
	"context"
	"log/slog"
	"time"

	runtime "github.com/eventgen/runtime-go"
)

type loggingAdapter struct {
	inner  runtime.TransportAdapter
	logger *slog.Logger
}

// WithLogging wraps an adapter with structured logging via slog.
func WithLogging(adapter runtime.TransportAdapter, logger *slog.Logger) runtime.TransportAdapter {
	if logger == nil {
		logger = slog.Default()
	}
	return &loggingAdapter{inner: adapter, logger: logger}
}

func (l *loggingAdapter) Publish(ctx context.Context, envelope runtime.EventEnvelope) (*runtime.PublishResult, error) {
	start := time.Now()
	l.logger.InfoContext(ctx, "publishing event",
		slog.String("eventId", envelope.EventID),
		slog.String("traceId", envelope.Metadata.TraceID),
	)
	result, err := l.inner.Publish(ctx, envelope)
	durationMs := time.Since(start).Milliseconds()
	if err != nil {
		l.logger.ErrorContext(ctx, "event publish failed",
			slog.String("eventId", envelope.EventID),
			slog.String("error", err.Error()),
			slog.Int64("durationMs", durationMs),
		)
		return nil, err
	}
	l.logger.InfoContext(ctx, "event published",
		slog.String("eventId", envelope.EventID),
		slog.String("messageId", result.MessageID),
		slog.Int64("durationMs", durationMs),
	)
	return result, nil
}
